ergotree-interpreter/src/sigma_protocol/prover.rs

```diff
@@ -892,24 +896,74 @@ fn step9_real_schnorr<P: Prover + ?Sized>(
             })
             .find(|prover_input| prover_input.public_image() == us.proposition)
         {
-            let z = dlog_protocol::interactive_prover::second_message(
-                priv_key,
-                us.randomness_opt.ok_or_else(|| {
-                    ProverError::Unexpected(format!("empty randomness in {:?}", us))
-                })?,
-                &challenge,
-            );
-            Ok(Some(
-                UncheckedSchnorr {
-                    proposition: us.proposition.clone(),
-                    commitment_opt: None,
-                    challenge,
-                    second_message: z,
-                }
-                .into(),
-            ))
+            let oc = hints_bag
```

Comment:
I suggest more idiomatic Rust approach to handling optional values. Consider the whole block inside `if let Some(challenge) ...` can be rewritten as:
```rust

        let z: SecondDlogProverMessage = if let Some(priv_key) = prover
            .secrets()
            .iter()
            .flat_map(|s| match s {
                PrivateInput::DlogProverInput(dl) => vec![dl],
                _ => vec![],
            })
            .find(|prover_input| prover_input.public_image() == us.proposition)
        {
            hints_bag
                .own_commitments()
                .into_iter()
                .find(|comm| comm.position == us.position)
                .map(|oc| {
                    dlog_protocol::interactive_prover::second_message(
                        priv_key,
                        oc.secret_randomness,
                        &challenge,
                    )
                })
                .unwrap_or({
                    dlog_protocol::interactive_prover::second_message(
                        priv_key,
                        us.randomness_opt.ok_or_else(|| {
                            ProverError::Unexpected(format!("empty randomness in {:?}", us))
                        })?,
                        &challenge,
                    )
                })
        } else {
            match hints_bag
                .real_proofs()
                .into_iter()
                .find(|comm| comm.position == us.position)
            {
                Some(tree) => {
                    let unchecked_tree = tree.unchecked_tree;
                    if let UncheckedTree::UncheckedLeaf(UncheckedLeaf::UncheckedSchnorr(
                        unchecked_schnorr,
                    )) = unchecked_tree
                    {
                        unchecked_schnorr.second_message
                    } else {
                        return Err(ProverError::SecretNotFound);
                    }
                }
                None => {
                    let bs = dlog_group::random_scalar_in_group_range(crypto_utils::secure_rng());
                    SecondDlogProverMessage { z: bs }
                }
            }
        };
        Ok(Some(
            UncheckedSchnorr {
                proposition: us.proposition,
                commitment_opt: None,
                challenge,
                second_message: z,
            }
            .into(),
        ))
```

Reply:
I rewrite the block same as dht case